### PR TITLE
WT-12328 Stop compacting if a file has grown in size between passes

### PR DESCRIPTION
**Pull request by Sean04**
The compact thread can get 'stuck' on a table for too long passing over it many times but failing to reclaim space due to other concurrent operations.

This adds an exit point to compact to stop iterating over a file when the file size grows between iterations.

### CODE REVIEW

File: src/include/block.h [hunk 1]
```diff
@@ -299,6 +299,7 @@ struct __wt_block {
     uint64_t compact_pages_rewritten_expected; /* The expected number of pages to rewrite */
     uint64_t compact_pages_skipped;            /* Pages skipped */
     uint32_t compact_session_id;               /* Session compacting */
+    wt_off_t compact_prev_size;                /* File size at the start of compact */
```
Comment (etienneptl) [inline src/include/block.h:302]: Please A -> Z

**Review by etienneptl — APPROVED** [review]
LGTM!
I wonder if we should declare the new field next to:
```
wt_off_t size; /* File size */
```

File: src/include/block.h (hunk 1 above)
Comment (etienneptl) [inline src/include/block.h:302]: And I suggest to change the comment:
```suggestion
    wt_off_t compact_prev_size;                /* File size at the start of a compaction pass */
```

**Comment by Sean04** [comment]
You could easily argue both ways but in my opinion, it should be with the compact attributes.

**Comment by etienneptl** [comment]
There are a few failures that need to be dealt with. Sean is OOO, I will have a look.

**Comment by pmacko86** [comment]
I like the overall approach here, but I have a question about the following corner case—can you please help me walk through it to make sure that this approach will handle it correctly?

Let's say that we have a file that in the first 80% has exactly the amount of free space for the last 20% on the file. After a compaction pass moves all the data, the first 80% of the file will be fully allocated, the last 20% of file will be in the "discard" extent list, and the "avail" list will be empty. The compaction will then take two checkpoint. Because the "avail" list is empty, the first checkpoint will actually grow the file to write its metadata, because there is no other space. The second checkpoint will be able to reuse the discarded space and write its metadata there. Under normal circumstances, the previous checkpoint metadata (written at the end of the file) will be no longer needed and will be discarded, but if the checkpoint retention is set to more than 1, this will not happen—and the file would grow between compaction passes despite successful compaction.

I think that we are not concerned with this case, because I *think* that the only way this could happen is if we would be skipping the next compaction phase anyways. Moreover, I don't think we expect customers to run with non-standard checkpoint retention, so maybe this issue is moot anyways.

Anyways, just thinking aloud. If this is not right and we do have to care about this corner case, let's discuss. Otherwise LGTM.

**Review by pmacko86 — APPROVED** [review]
LGTM.

**Comment by etienneptl** [comment]
> Let's say that we have a file that in the first 80% has exactly the amount of free space for the last 20% on the file. After a compaction pass moves all the data, the first 80% of the file will be fully allocated, the last 20% of file will be in the "discard" extent list, and the "avail" list will be empty. The compaction will then take two checkpoint. Because the "avail" list is empty, the first checkpoint will actually grow the file to write its metadata, because there is no other space. The second checkpoint will be able to reuse the discarded space and write its metadata there. Under normal circumstances, the previous checkpoint metadata (written at the end of the file) will be no longer needed and will be discarded, but if the checkpoint retention is set to more than 1, this will not happen—and the file would grow between compaction passes despite successful compaction.
> 
> I think that we are not concerned with this case, because I _think_ that the only way this could happen is if we would be skipping the next compaction phase anyways. Moreover, I don't think we expect customers to run with non-standard checkpoint retention, so maybe this issue is moot anyways.

@pmacko86, thank you for sharing this scenario, it describes a potential problem. If the file has grown after the two checkpoints, we would detect the file growth and skip compaction. It would be an improvement to detect where the growth comes from but it's too hard as of now, especially if there is parallel activity.
Regarding checkpoint retention, it's part of the `debug_mode` configuration list - it's a fair point, but I don't think we should look too much into it.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Pull request merged**